File: python/day04.py
from aoc import *


def part_1(data):
    return sum((a <= c and d <= b) or (c <= a and b <= d)
               for a, b, c, d in data)


def part_2(data):
    return sum(not(a > d or b < c) for a, b, c, d in data)


def solve(file=4):
    data = [integers(line, negative=False) for line in read_input(file)]
    return part_1(data), part_2(data)


print(solve())










def test():
    assert solve("04_test") == (2, 4)
    assert solve() == (569, 936)

File: clojure/day04.clj
(ns day04
  (:require aoc))


(defn fully-contain? [[a b c d]]
  (or (<= a c d b)
      (<= c a b d)))

(defn overlap? [[a b c d]]
  (or (<= c a d)
      (<= a c b)))

(defn parse-input [input]
  (aoc/parse-input input #(aoc/integers % {:negative? false})))

(defn solve
  ([] (solve (aoc/read-file 4)))
  ([input]
   (let [assignments-pairs (parse-input input)]
     [(aoc/count-if fully-contain? assignments-pairs)
      (aoc/count-if overlap?       assignments-pairs)])))


(solve)
